market_maker_keeper/gas.py

```diff
@@ -49,12 +82,59 @@ def get_gas_price(self, time_elapsed: int) -> Optional[int]:
                                       max_price=100*self.GWEI).get_gas_price(time_elapsed)
 
 
+class DynamicGasPrice(NodeAwareGasPrice):
+    every_secs = 42
+
+    def __init__(self, web3: Web3, arguments: Namespace):
+        assert isinstance(web3, Web3)
+
+        self.gas_station = None
+        self.fixed_gas = None
+        self.web3 = web3
+        if arguments.oracle_gas_price:
+            self.gas_station = Aggregator(refresh_interval=60, expiry=600,
+                                          ethgasstation_api_key=arguments.ethgasstation_api_key,
+                                          poa_network_alt_url=arguments.poanetwork_url,
+                                          etherscan_api_key=arguments.etherscan_api_key,
+                                          gasnow_app_name="makerdao/market-maker-keeper")
+        elif arguments.fixed_gas_price:
+            self.fixed_gas = int(round(arguments.fixed_gas_price * self.GWEI))
+        self.initial_multiplier = arguments.gas_initial_multiplier
+        self.reactive_multiplier = arguments.gas_reactive_multiplier
+        self.gas_maximum = int(round(arguments.gas_maximum * self.GWEI))
+        if self.fixed_gas:
+            assert self.fixed_gas <= self.gas_maximum
+
+    def __del__(self):
+        if self.gas_station:
+            self.gas_station.running = False
+
+
+    def get_gas_price(self, time_elapsed: int) -> Optional[int]:
+        # start with fast price from the configured gas API
+        fast_price = self.gas_station.fast_price() if self.gas_station else None
+
+        # if API produces no price, or remote feed not configured, start with a fixed price
+        if fast_price is None:
+            if self.fixed_gas:
+                initial_price = self.fixed_gas
+            else:
+                initial_price = int(round(self.get_node_gas_price() * self.initial_multiplier))
+        # otherwise, use the API's fast price, adjusted by a coefficient, as our starting point
+        else:
+            initial_price = int(round(fast_price * self.initial_multiplier))
+
+        return GeometricGasPrice(initial_price=initial_price,
+                                 every_secs=DynamicGasPrice.every_secs,
+                                 coefficient=self.reactive_multiplier,
+                                 max_price=self.gas_maximum).get_gas_price(time_elapsed)
+
 class GasPriceFactory:
     @staticmethod
-    def create_gas_price(arguments) -> GasPrice:
+    def create_gas_price(web3: Web3, arguments: Namespace) -> GasPrice:
         if arguments.smart_gas_price:
-            return SmartGasPrice(arguments.ethgasstation_api_key)
-        elif arguments.gas_price:
-            return FixedGasPrice(arguments.gas_price)
+            return SmartGasPrice(web3, arguments)
```

Comment:
Legacy `SmartGasPrice` still just takes the EGS api key as an arg; not sure why we're passing the entire collection of keeper args now.

Reply:
Fixed to just pass `arguments.ethgasstation_api_key` to `SmartGasPrice`.